modules/gapi/src/streaming/vpl/surface/surface.cpp

```diff
@@ -0,0 +1,74 @@
+#include "streaming/vpl/surface/surface.hpp"
+#include "logger.hpp"
+
+#ifdef HAVE_ONEVPL
+
+namespace cv {
+namespace gapi {
+namespace wip {
+
+Surface::Surface(std::unique_ptr<handle_t>&& surf, std::shared_ptr<void> associated_memory) :
+    workspace_memory_ptr(associated_memory),
+    mfx_surface(std::move(surf)),
+    mirrored_locked_count() {
+
+    GAPI_Assert(mfx_surface && "Surface is nullptr");
+    mirrored_locked_count.store(mfx_surface->Data.Locked);
+    GAPI_LOG_DEBUG(nullptr, "create surface: " << mfx_surface <<
+                            ", locked count: " << mfx_surface->Data.Locked);
+}
+
+Surface::~Surface() {
+    GAPI_LOG_DEBUG(nullptr, "destroy surface: " << mfx_surface <<
+                            ", worspace memory counter: " << workspace_memory_ptr.use_count());
+}
+
+std::shared_ptr<Surface> Surface::create_surface(std::unique_ptr<handle_t>&& surf,
+                                                 std::shared_ptr<void> accociated_memory) {
+    surface_ptr_t ret {new Surface(std::move(surf), accociated_memory)};
+    return ret;
+}
+
+std::shared_ptr<Surface> Surface::get_ptr() {
+    return shared_from_this();
+}
+
+Surface::handle_t* Surface::get_handle() const {
+    return mfx_surface.get();
+}
+
+const Surface::info_t& Surface::get_info() const {
+    return mfx_surface->Info;
+}
+
+const Surface::data_t& Surface::get_data() const {
+    return mfx_surface->Data;
+}
+
+size_t Surface::get_locks_count() const {
+    return mirrored_locked_count.load();
```

Comment:
Not sure about this method. `mirrored_locked_count` is an atomic but it can be changed while you call `get_locks_count`

Reply:
yeah, it can be INCREMENTED for this  surface BEFORE and not AFTER.

ThreadA: increment() it  and asks for load() in this method
ThreadB: **de**crement this counter only

So this method checks availability: we could set one surface counter (it's busy for example) and then we would check second surface on lock counter in the same thread (let's assume it is free). meantime the first surface may be releases in another thread simultaneously. But it's OK because we just need to get any instance of free surface ( second in example). And this method doesn't suggest "busy wait" - it's is responsibility of upper-level code